### PR TITLE
[Core] Support negative conditions in SDXL

### PR DESCRIPTION
**Pull request by sayakpaul**
Fixes: https://github.com/huggingface/diffusers/issues/4208. 

I don't think we need to immediately support dropping out embeddings as it only affects fine-tuning workflows. 

Testing code:

```python
from diffusers import StableDiffusionXLPipeline
import torch

pipe = StableDiffusionXLPipeline.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.float16, variant="fp16", use_safetensors=True
).to("cuda")

prompt = "Astronaut in a jungle, cold color palette, muted colors, detailed, 8k"
generator = torch.manual_seed(1000)
negative_conditions = {
    "original_size": (512, 512),
    "target_size": (1024, 1024),
    "crops_coords_top_left": (0, 0)
}

image_with_neg_cond = pipe(
    prompt=prompt, generator=generator, num_inference_steps=25, negative_conditions=negative_conditions
).images[0]
image_with_neg_cond
```

Comparison (left: without any neg condition, right: with neg condition):

neg_cond_comparison

It also allows us to see the influence of `original_size` on `negative_conditions`: 

```python
prompt = "Astronaut in a jungle, cold color palette, muted colors, detailed, 8k"

negative_conditions = {
    "original_size": (512, 512),
    "target_size": (1024, 1024),
    "crops_coords_top_left": (0, 0)
}

original_sizes_to_test = [(128, 128), (256, 256), (512, 512)]
test_images = []

for original_size in original_sizes_to_test:
    negative_conditions.update({"original_size": original_size})
    generator = torch.manual_seed(1000)
    image_with_neg_cond = pipe(
        prompt=prompt, generator=generator, num_inference_steps=25, negative_conditions=negative_conditions
    ).images[0]
    test_images.append(image_with_neg_cond)
```

different_original_sizes

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:610]: Even though we already have lots of function input arguments here, I would tend towards doing the following:
```suggestion
        negative_original_size: Optional[Tuple[int, int]] = None,
        negative_crops_coords_top_left: Tuple[int, int] = (0, 0),
        negative_target_size: Optional[Tuple[int, int]] = None,
```

Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:610]: I think this is more consistent

**Review by patrickvonplaten — COMMENTED** [review]
Great job an getting this working! I agree with your judgement that randomly dropping embeddings is not (yet) needed for fine-tuning as I'd also think that this feature was mostly used for pre-training.

Could you maybe also apply this nice improvement to all other SDXL pipelines (Img2Img, ControlNet, Inpainting)?

**Comment by sayakpaul** [comment]
@stevhliu could you also review the doc related changes?

File: docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_xl.md
Comment (sayakpaul) [inline docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_xl.md:26]:
```suggestion
- One can make use of `negative_original_size`, `negative_crops_coords_top_left`, and `negative_target_size` to influence the generation process.
```

**Review by patrickvonplaten — APPROVED** [review]
Clean! Great job

**Comment by sayakpaul** [comment]
Will merge after Steven's go ahead on the docs as the doc I am changing is very highly regarded.

File: docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_xl.md
Comment (stevhliu) [inline docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_xl.md:78]: This feels a bit vague, and I think it could be more explicit in explaining what passing `negative_original_size`, `negative_target_size` and `negative_crops_coords_top_left` do. I think it'd be more impactful if it is immediately clear when the user reads it (ie how do these negative conditions affect generation?).

Comment (sayakpaul) [reply docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_xl.md:78]: It's similar to how classifier guidance is done. With these negative conditions, the model tends to move away from these and focus on non negative conditions instead. Feel free to suggest a text that you think is best suitable here.

Comment (stevhliu) [reply docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_xl.md:78]: Maybe something like:

"You can additionally pass *negative conditions* about an image's size and position to avoid undesirable cropping behavior in the generated image, and improve image resolution. "

Comment (sayakpaul) [reply docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_xl.md:78]: Perfect! Will use it and resolve the conflicts before merging. Thanks a lot!

**Comment by yiyixuxu** [comment]
@stevhliu 

we definitely need some doc for these micro-conditions now 😬 
It was confusing to me and getting more every day

**Comment by stevhliu** [comment]
For sure, I'll update https://github.com/huggingface/diffusers/pull/4428 with a new section about the micro-conditions!

**Pull request merged**
